### PR TITLE
#2034 D&D and Copy-Paste disable Mix of Mode and State in Hierarchy

### LINKED ISSUE #2034: D&D disable Mix of Mode and State in Hierarchy
Drag and drop and Copy-Paste of region should not be possible in statemachine with modes in it. EXCEPT IF the related existing preference inside Capella/Model is used, i.e. mix M & S allowed

### PR DESCRIPTION
**Pull request by MalinaStefaniaStoicanescu**
- [x] drag and drop of region should not be possible in statemachine with modes in it.
- [x]  update JUnits

Fixes #2034 
Change-Id: I05f9d72e26af0406d34f1f1b0d2392497b9306d9
Signed-off-by: MalinaStefaniaStoicanescu <malina.stoicanescu@thalesgroup.com>

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java [hunk 1]
```diff
@@ -335,6 +350,183 @@ protected boolean checkCompatibility(List<EObject> modelElements, EList<EReferen
...
+  }
+
+  /**
+   * Get all states of a region
+   * 
+   * @param region
+   *          Region
+   * @return all states of a region
+   */
+  private EList<AbstractState> getAllRegionStates(Region region) {
```
Comment (minhtutonthat) [inline core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java:387]: org.polarsys.capella.core.model.helpers.move.MoveHelper.getAllModeState(Region) does not work?

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java [hunk 2]
```diff
@@ -335,6 +350,183 @@ protected boolean checkCompatibility(List<EObject> modelElements, EList<EReferen
...
+  }
+
+  /**
+   * Get the first region (the "root" region = first region under StateMachine) of a State or a Region
+   * 
+   * @param object
+   *          Region or State
+   * @return the root region or null
+   */
+  public Region getRootRegion(EObject object) {
```
Comment (minhtutonthat) [inline core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java:415]: Methods like this merit to be put in a helper (e.g. StateMachineExt).

**Review by minhtutonthat — CHANGES_REQUESTED** [review]

**Comment by minhtutonthat** [comment]
Could you also add "Fixes #issuenumber" in your commit message so that the issue can be linked to this PR?

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java [hunk 3]
```diff
@@ -335,6 +350,183 @@ protected boolean checkCompatibility(List<EObject> modelElements, EList<EReferen
...
+
+  /**
+   * Depending on mixed hierarchy mode state preference, determine if we can move a region into a State Machine
+   * @param sourceRegion
+   *          The region to be moved
+   * @param stateMachine
+   *          The state machine where the region should be moved
+   * @return true if the region could be moved, false if not
+   */
+  public boolean canMoveRegion(Region sourceRegion, StateMachine stateMachine) {
```
[the remark this answers is not in the document]
Comment (MalinaStefaniaStoicanescu) [reply core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java:362]: done

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java [hunk 4]
```diff
@@ -169,6 +171,19 @@ public IStatus checkSemanticRules(List<EObject> selectedElements, EObject target
...
+        } else if (source instanceof Region) {
+          if (target instanceof AbstractState) {
+            isOK = canMoveRegion((Region) source, (AbstractState) target);
+          }
+
+          if (target instanceof StateMachine) {
+            isOK = canMoveRegion((Region) source, (StateMachine) target);
+          }
+
+          if (target instanceof Region) {
```
[the remark this answers is not in the document]
Comment (MalinaStefaniaStoicanescu) [reply core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java:185]: for copy paste no, but for drag and drop i think it's ok this check.

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java [hunk 5]
```diff
@@ -335,6 +350,183 @@ protected boolean checkCompatibility(List<EObject> modelElements, EList<EReferen
...
+  /**
+   * Populate 'states' parameter with all states of a region
+   * 
+   * @param region
+   *          Region
+   * @param states
+   *          List with all encountered states - at the beginning is empty - it will be updated by appending all the
+   *          states of the region
+   */
+  private void populateListWithAllRegionStates(Region region, EList<AbstractState> states) {
```
[the remark this answers is not in the document]
Comment (MalinaStefaniaStoicanescu) [reply core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java]: this method was removed

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java (hunk 1 above)
Comment (MalinaStefaniaStoicanescu) [reply core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java]: you are right, i deleted it

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java (hunk 2 above)
Comment (MalinaStefaniaStoicanescu) [reply core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java]: moved :)

File: core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (minhtutonthat) [reply core/plugins/org.polarsys.capella.core.model.helpers/src/org/polarsys/capella/core/model/helpers/move/MoveHelper.java:185]: I meant in the meta-model it's not possible for a Region to contain another Region. Could you check that without this check, the D&D Region to Region is still not impossible? If that's the case, no need to check.

**Review by minhtutonthat — CHANGES_REQUESTED** [review]
Could you also add JUnit tests to cover different cases? You could insprire from org.polarsys.capella.test.model.ju.dnd.DnDTestSuite.

**Review by minhtutonthat — APPROVED** [review]

**Pull request merged**